From 1924a8ee1f9d654a79dc37354a66536c24295368 Mon Sep 17 00:00:00 2001
From: Rich Bono <rbono@redhat.com>
Date: Thu, 20 Nov 2014 21:23:26 -0500
Subject: [scsi] pm80xx: fix problem of pm8001_work_fn reseting incorrect phy device

Message-id: <1416518623-7080-6-git-send-email-rbono@redhat.com>
Patchwork-id: 100406
O-Subject: [RHEL7.1 e-stor PATCH V3 05/22] [SCSI] pm80xx: fix problem of pm8001_work_fn reseting incorrect phy device
Bugzilla: 1110943
RH-Acked-by: Rob Evers <revers@redhat.com>
RH-Acked-by: Tomas Henzl <thenzl@redhat.com>

From: XinHong Zhu <zxh3737@163.com>

If a phy device is removed, the device can get error of I/O and HBA maybe
receieve IO_OPEN_CNX_ERROR_IT_NEXUS_LOSS of event which causes pm8001_work_fn
to reset the phy device but in pm8001_task_exec we don't assign a value for
the device field of the ccb and in other case a ccb used have device field
set, when ccb is freed the field device of the ccb don't be set NULL.So there
is possibility of getting another device reset in function mpi_ssp_completion.

Signed-off-by: zhuxh <zxh3737@163.com>
Acked-by: Lindar Liu <lindar_liu@usish.com>
Signed-off-by: James Bottomley <JBottomley@Parallels.com>
(cherry picked from commit aed97b8b8194e818a94cc5ce5662f19b8bdd8adc)

Signed-off-by: Rich Bono <rbono@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/pm8001/pm8001_sas.c b/drivers/scsi/pm8001/pm8001_sas.c
index ba3088c..8a44bc9 100644
--- a/drivers/scsi/pm8001/pm8001_sas.c
+++ b/drivers/scsi/pm8001/pm8001_sas.c
@@ -434,6 +434,7 @@ static int pm8001_task_exec(struct sas_task *task, const int num,
   ccb->n_elem = n_elem;
   ccb->ccb_tag = tag;
   ccb->task = t;
+  ccb->device = pm8001_dev;
   switch (t->task_proto) {
   case SAS_PROTOCOL_SMP:
    rc = pm8001_task_prep_smp(pm8001_ha, ccb);
-- 
1.7.1